File: Leetcode_Hot_100/code/golang/02_add-two-numbers.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func addNode(iter *ListNode, val int) (node *ListNode) {
	node = new(ListNode)
	node.Val = val
	node.Next = nil
	if iter != nil {
		iter.Next = node
	}
	return
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	a, b := l1.Val, l2.Val
	l1, l2 = l1.Next, l2.Next
	sum := a + b
	carry := sum / 10
	head := addNode(nil, sum%10)
	iter := head
	for !(l1 == nil && l2 == nil) {
		a = 0
		if l1 != nil {
			a = l1.Val
			l1 = l1.Next
		}
		b = 0
		if l2 != nil {
			b = l2.Val
			l2 = l2.Next
		}
		sum = a + b + carry
		iter = addNode(iter, sum%10)
		carry = sum / 10
	}
	if carry != 0 {
		addNode(iter, carry)
	}
	return head
}

func main() {

}

File: Leetcode_Hot_100/code/cpp/02_add-two-numbers.cpp
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
private:
    ListNode* addNode(ListNode* iter, int value)
    {
        ListNode* node = new(ListNode);
        node->val = value;
        node->next = nullptr;
        if (iter != nullptr)
            iter->next = node;
        return node;
    }
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int a = l1->val, b = l2->val;
        int value = (a + b) % 10;
        int offset = (a + b) / 10;
        l1 = l1->next;
        l2 = l2->next;
        ListNode* head = addNode(nullptr, value);
        ListNode* iter = head;

        while(!(l1 == nullptr && l2 == nullptr))
        {
            a = 0;
            if (l1 != nullptr)
            {
                a = l1->val;
                l1 = l1->next;
            }
            b = 0;
            if (l2 != nullptr)
            {
                b = l2->val;
                l2 = l2->next;
            }
            value = (a + b + offset) % 10;
            offset = (a + b + offset) / 10;
            iter = addNode(iter, value);
        }
        if (offset)
            iter = addNode(iter, offset);
        return head;
    }
};